src/App.tsx

```diff
@@ -1,24 +1,345 @@
+
+  const inputRef = useRef<HTMLInputElement>(null);
+
+  const [
+    currentError,
+    setCurrentError,
+  ] = useState<ErrorType>(ErrorType.NoError);
+
+  const timerId = useRef<NodeJS.Timeout>();
+
+  const onError = (error: ErrorType) => {
+    if (timerId.current) {
+      clearTimeout(timerId.current);
+    }
+
+    setCurrentError(error);
+    timerId.current = setTimeout(
+      () => setCurrentError(ErrorType.NoError),
+      3000,
+    );
+  };
+
+  const handleSubmit = async (
+    event: FormEvent<HTMLFormElement>,
+    newTodoText: string,
+    setNewTodoText: Dispatch<SetStateAction<string>>,
+  ) => {
+    event.preventDefault();
+
+    if (!newTodoText.trim()) {
+      onError(ErrorType.EmptyTitle);
+
+      return;
+    }
+
+    const newTodo = {
+      id: 0,
+      userId: USER_ID,
+      title: newTodoText.trim(),
+      completed: false,
+    };
+
+    try {
+      setTempTodo(() => newTodo);
+      await addTodo(newTodo)
+        .then((todo) => setTodosFromServer(
+          prevTodos => ([...prevTodos, todo]),
+        ))
+        .then(() => {
+          setNewTodoText('');
+        });
+    } catch {
+      onError(ErrorType.UnableToAdd);
+    } finally {
+      setTempTodo(() => null);
+
+      setTimeout(() => {
+        if (inputRef.current) {
+          inputRef.current.focus();
+        }
+      }, 100);
+    }
+  };
+
+  const handleSingleTodoDelition = async (todoId: number) => {
+    try {
+      setSelectedTodoIds(prevIds => [...prevIds, todoId]);
+
+      await deleteTodo(todoId).then(() => {
+        setTodosFromServer(
+          prevTodos => prevTodos.filter(todo => todo.id !== todoId),
+        );
+      });
+    } catch {
+      onError(ErrorType.UnableToDelete);
+    } finally {
+      setTimeout(
+        () => {
+          setSelectedTodoIds(ids => {
+            const idsCopy = [...ids];
+
+            idsCopy.splice(ids.indexOf(todoId), 1);
+
+            return idsCopy;
+          });
+
+          if (inputRef.current) {
+            inputRef.current.focus();
+          }
+        },
+        300,
+      );
+    }
+  };
+
+  const completedTodoIds = todosFromServer
+    .filter(todo => todo.completed)
+    .map(todo => todo.id);
+
+  const handleCompletedTodosDelition = async () => {
+    const processedIds: number[] = [];
+
+    try {
+      setSelectedTodoIds(prev => [...prev, ...completedTodoIds]);
+
+      completedTodoIds.forEach(async (todoId) => {
+        await deleteTodo(todoId).catch(() => onError(ErrorType.UnableToDelete))
+          .then(() => {
+            processedIds.push(todoId);
+          });
+      });
+    } catch {
+      onError(ErrorType.UnableToDelete);
+    } finally {
+      setTimeout(
+        () => {
+          setSelectedTodoIds(
+            ids => ids.filter(id => !completedTodoIds.includes(id)),
+          );
+
+          setTodosFromServer(
+            todos => todos.filter(todo => !processedIds.includes(todo.id)),
+          );
```

Comment:
Remove todos from an array you should do only in positive case instaed of finally method

Reply:
> Remove todos from an array you should do only in positive case instaed of finally method

Not fixed